File: kfch/src/main/scala/xyz/sigmalab/kfch/util/KafkaUtil.scala
package xyz.sigmalab.kfch.util

import java.util
import java.util.{Collections, Properties}

import org.apache.kafka.clients.admin.{AdminClient, AdminClientConfig}
import org.apache.kafka.clients.consumer.{Consumer, ConsumerConfig, KafkaConsumer}
import org.apache.kafka.clients.producer.{KafkaProducer, Producer, ProducerConfig}
import org.apache.kafka.common.KafkaFuture
import org.apache.kafka.common.serialization.{StringDeserializer, StringSerializer}

import scala.concurrent.{Await, Future, Promise}
import scala.concurrent.duration._
import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer

object KafkaUtil {

    def adminClient(server: String) = {
        val conf = {
            val p = new Properties()
            p.put(AdminClientConfig.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG, 2000) // 2.7.0
            // m.put(AdminClientConfig.SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG, 2000) // 2.7.0
            p.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, server)
            p.put(AdminClientConfig.CLIENT_ID_CONFIG, "ADMIN_CLIENT")
            p
        }
        AdminClient.create(conf)
    }

    def newProducer(server: String): Producer[String, String] = {
        val props = {
            val p = new Properties()
            p.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, server)
            p.put(ProducerConfig.RETRIES_CONFIG, 0)
            p.put(ProducerConfig.ACKS_CONFIG, "all")
            p.put(ProducerConfig.CLIENT_ID_CONFIG, "PRODUCER_CLIENT")
            p
        }
        new KafkaProducer(props, new StringSerializer, new StringSerializer)
    }

    /**
     *
     * ERRORS: enable.auto.commit cannot be set to true when default group id (null) is used
     *
     * @param server
     * @param autoCommitIntervals
     * @param offsetReset "earliest" | "latest" | "none"
     * @return
     */
    def newConsumer(
        server: String,
        groupId: String,
        topic: String,
        autoCommitIntervals: Int = 1000,
        offsetReset: String = "earliest"
    ): Consumer[String, String] = {
        val props = {

            val p = new Properties()
            p.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, server)

            if (autoCommitIntervals > 0) {
                p.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, true)
                p.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, autoCommitIntervals)
            } else p.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)

            p.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, offsetReset)
            p.put(ConsumerConfig.GROUP_ID_CONFIG, groupId)

            p.put(ConsumerConfig.CLIENT_ID_CONFIG, "CONSUMER_CLIENT")
            p
        }
        val consumer = new KafkaConsumer(props, new StringDeserializer, new StringDeserializer)
        consumer.subscribe(Collections.singleton(topic))
        consumer
    }

    implicit class KafkaFtOpt[T](val self: KafkaFuture[T])
        extends AnyVal {
        def toScalaFuture: Future[T] = {
            val p = Promise[T]
            self.whenComplete { (t, error) =>
                if (error == null) p.success(t)
                else p.failure(error)
            }
            p.future
        }

        def toft: Future[T] = toScalaFuture

        def result: T = Await.result(toft, 30.seconds)

        def getIter[X](implicit ev: T <:< java.util.Collection[X]): Iterable[X] = {
            result.asInstanceOf[util.Collection[X]].asScala
        }

        def getMap[A, B](implicit ev: T <:< java.util.Map[A, B]): Map[A, B] = {
            result.asInstanceOf[util.Map[A, B]].asScala.toMap
        }
    }

    def consumeN[K, V](consumer: Consumer[K, V], n: Int, maxWait: Duration): Iterable[(K, V)] = {
        val jtimeout = java.time.Duration.ofMillis(maxWait.toMillis)
        val start = System.currentTimeMillis();
        val expectedEnd = start + maxWait.toMillis
        var cntu = true;
        var required = n;
        val buf = new ArrayBuffer[(K, V)](n)

        do {
            consumer.poll(jtimeout) match {
                case batch if batch.isEmpty =>
                    println(s"PolledBatch: EMPTY")
                case batch =>
                    println(s"PolledBatch: ${batch.count()}")
                    val takeN = if (batch.count() <= required) batch.count() else required
                    buf.appendAll(batch.iterator().asScala.take(takeN).map { i => i.key -> i.value })
                    required -= takeN
            }

            if (required == 0) cntu = false
            val diff = expectedEnd - System.currentTimeMillis()
            if (diff < 0) cntu = false
            println(s"Continue? ${required} / ${diff}")

        } while (cntu);

        buf
    }

    val NO_OFFSET_STRATEGY_EARLIEST = "earliest"
    val NO_OFFSET_STRATEGY_LATEST = "latest"
    val NO_OFFSET_STRATEGY_FAIL = "none"


    def debugOffsetAndConsumeN[K, V](
        server: String, topic: String, group: String,
        n: Int, maxWait: Duration,
        noOffsetStrategy: String = NO_OFFSET_STRATEGY_EARLIEST
    ) = {
        println("Connecting ...")
        val admin = adminClient(server)
        val consumer = newConsumer(server, group, topic, -1, noOffsetStrategy)

        println(s"Fetching Offsets of '${group}'")
        val offsets =
            admin.listConsumerGroupOffsets(group)
                .partitionsToOffsetAndMetadata()
                .getMap



        if (offsets.isEmpty)
            println(s"NO OFFSET, Strategy: ${noOffsetStrategy}")
        else {
            offsets.foreach { case (k, v) => println(s"$k => $v") }
            println("Deleting Offsets ...")
            admin.deleteConsumerGroupOffsets(group, offsets.map(_._1).toSet.asJava).all().result
        }

        println(s"NO_OFFSET_STRATEGY: ${noOffsetStrategy}")

        consumeN(consumer, n, maxWait).foreach { i => println(s" :: $i")}
        consumer.commitSync()

        println(s"Fetching Offsets of '${group}'")
        val newOffsets =
            admin.listConsumerGroupOffsets(group)
                .partitionsToOffsetAndMetadata()
                .getMap

        if (newOffsets.isEmpty) println("NO OFFSET !")
        else newOffsets.foreach { case (k, v) => println(s"$k => $v") }
    }

}

File: ksutil/src/main/java/etc/KafkaUtil.java
package etc;

import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.AdminClientConfig;
import org.apache.kafka.clients.admin.Config;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.clients.consumer.*;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.Producer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;

import java.time.Duration;
import java.util.*;

public class KafkaUtil {

    public static class PropsBuilder {
        private final Properties props;

        public PropsBuilder(Properties props) {
            this.props = props;
        }

        public PropsBuilder() {
            this(new Properties());
        }

        public PropsBuilder put(String k, Object v) {
            props.put(k, v.toString());
            return this;
        }

        public Object get(String k) {
            return props.get(k);
        }

        public Properties result() {
            return props;
        }
    }

    public static PropsBuilder props() {
        return new PropsBuilder();
    }

    public static enum NoOffsetStrategy {
        EARLIEST("earliest"),
        LATEST("latest"),
        ERROR("none");

        private final String value;

        NoOffsetStrategy(String value) {
            this.value = value;
        }

        public String getV() {
            return value;
        }
    }

    private final String name;
    private final String bootstraperServer;

    private final AdminClient admin;

    public KafkaUtil(String name, String bootstraperServer) {
        this.name = name;
        this.bootstraperServer = bootstraperServer;
        this.admin = AdminClient.create(propsAdminClient().result());
    }

    public String newName(String s) {
        return String.format("%s_%s_%d", name, s, System.currentTimeMillis());
    }

    public String getBootstraperServer() {
        return bootstraperServer;
    }

    public String getName() {
        return name;
    }

    protected PropsBuilder propsAdminClient() {
        return props()
                .put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstraperServer)
                .put(AdminClientConfig.CLIENT_ID_CONFIG, newName("kafkaadmin"))
                .put(AdminClientConfig.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG, 5000);
    }

    protected PropsBuilder propsProducer() {
        return props()
                .put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstraperServer)
                .put(ProducerConfig.RETRIES_CONFIG, 0)
                .put(ProducerConfig.ACKS_CONFIG, "all")
                .put(ProducerConfig.CLIENT_ID_CONFIG, newName("kafkaproducer"));
    }

    public void publish(String topic, List<String> messages) {
        try {
            Producer<String, String> producer =
                    new KafkaProducer<String, String>(propsProducer().result(),
                            new StringSerializer(), new StringSerializer());

            for (String i : messages) {
                producer.send(new ProducerRecord<>(topic, i));
            }
            producer.flush();
            producer.close();
        } catch (Exception cause) {
            throw new RuntimeException(cause);
        }
    }


    protected PropsBuilder propsConsumer() {
        PropsBuilder p = props();
        p.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstraperServer);
        p.put(ConsumerConfig.CLIENT_ID_CONFIG, newName("kafkaconsumer"));
        p.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
        return p;
    }

    public List<String> consumeN(String topic, String group, int n, Duration maxWait,
                                 NoOffsetStrategy nofs) {
        try {
            Properties props =
                    propsConsumer()
                            .put(ConsumerConfig.GROUP_ID_CONFIG, group)
                            .put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, nofs.getV())
                            .result();

            Consumer<String, String> consumer =
                    new KafkaConsumer<String, String>(props, new StringDeserializer(), new StringDeserializer());

            consumer.subscribe(Collections.singleton(topic));



            ArrayList<String> buf = new ArrayList<>(n);
            int required = n;
            boolean cntu = true;
            long start = System.currentTimeMillis();
            long expectedEnd = start + maxWait.toMillis();

            do {
                Duration timeout =
                        Duration.ofMillis(expectedEnd - System.currentTimeMillis());

                ConsumerRecords<String, String> records =
                        consumer.poll(timeout);

                for (ConsumerRecord<String, String> r : records) {
                    buf.add(r.value());
                    required--;
                    if (required <= 0) break;
                }

                if (required <= 0) cntu = false;
                long timeDiff = expectedEnd - System.currentTimeMillis();
                if (timeDiff <= 0) cntu = false;

            } while (cntu);

            consumer.commitSync();
            consumer.close();

            return buf;
        } catch (Exception cause) {
            throw new RuntimeException(cause);
        }
    }

    public Config createTopic(String name) {
        try {
            return admin.createTopics(Collections.singleton(new NewTopic(name, 1,(short) 1))).config(name).get();
        } catch (Exception cause) {
            throw new RuntimeException(cause);
        }
    }

    public List<String> genAndPublish(String topic, String format, int start, int stop, int step) {
        ArrayList<String> buf = new ArrayList<>();
        for (int i=start; i < stop; i+=step)
            buf.add(String.format(format, i));
        publish(topic, buf);
        return buf;
    }

    public Map<TopicPartition, OffsetAndMetadata> getOffsetsOfGroup(String group) {
        try {
            return admin.listConsumerGroupOffsets(group).partitionsToOffsetAndMetadata().get();
        } catch (Exception cause) {
            throw new RuntimeException(cause);
        }
    }
}
